Minor improvements to `rustc_ast_passes`

nnethercote opened the pull request:
Some improvements I found while looking at this code.

r? @compiler-errors

nnethercote committed:
Streamline `gate_feature_*` macros.

The debug probably isn't useful, and assigning all the `$foo`
metavariables to `foo` variables is verbose and weird. Also, `$x:expr`
usually doesn't have a space after the `:`.

nnethercote committed:
Cover two more cases in the `gate_doc` macro.

nnethercote committed:
Use a slice pattern to neaten a condition.

nnethercote committed:
Use `if let` to reduce some excessive indentation.

compiler-errors commented on compiler/rustc_ast_passes/src/feature_gate.rs:
```diff
@@ -144,15 +144,11 @@ impl<'a> PostExpansionVisitor<'a> {
-                crate::fluent_generated::ast_passes_forbidden_non_lifetime_param,
-            )
-            .emit();
+        if !non_lt_param_spans.is_empty() {
```
What this will do is emit a feature error for every single span. What I want is for it to be a single feature error with multiple primary spans.

I prefer if this is not changed, or it is fixed in the right way.

compiler-errors replied:
IDK why there isn't a test for this. Here's one:

```
fn test() -> for<T, S> fn(T, S) { todo!() }
```

```
error[E0658]: only lifetime parameters can be used in this context
 --> src/lib.rs:1:18
  |
1 | fn test() -> for<T, S> fn(T, S) { todo!() }
  |                  ^  ^
  |
  = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
```

nnethercote replied:
:+1: 

I have updated the code, removing the invalid commit, and adding three new commits: another small refactor, an alternative fix for the FIXME, and a test update. Hopefully that's closer to what is necessary :)  Those macros still annoy me, but I can't see a better way to do things without increasing the amount of boilerplate.

nnethercote committed:
Rearrange the `gate_feature_*` macros.

nnethercote committed:
Fix a `FIXME`, by adding a `gate_multi` macro.

Note that this adds the `span.allows_unstable` checking that this case
previously lacked.

nnethercote committed:
Test the multispan case in `tests.ui/bounds-lifetime.rs`.

nnethercote commented:
I fixed the conflicts.

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 5b391b01ce66240bfcdbd92ac45339f26db283c2 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 5b391b01ce66240bfcdbd92ac45339f26db283c2 -->
<!-- homu: {"type":"Approved","sha":"5b391b01ce66240bfcdbd92ac45339f26db283c2","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5b391b01ce66240bfcdbd92ac45339f26db283c2 with merge 62270fb4d674fa109148c3a2618e4db9e03cd91c...
<!-- homu: {"type":"BuildStarted","head_sha":"5b391b01ce66240bfcdbd92ac45339f26db283c2","merge_sha":"62270fb4d674fa109148c3a2618e4db9e03cd91c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6731081708/job/18299083489)
Approved by: compiler-errors
Pushing 62270fb4d674fa109148c3a2618e4db9e03cd91c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6731081708/job/18299083489"},"merge_sha":"62270fb4d674fa109148c3a2618e4db9e03cd91c"} -->

rust-timer commented:
Finished benchmarking commit (62270fb4d674fa109148c3a2618e4db9e03cd91c): [comparison URL](https://perf.rust-lang.org/compare.html?start=46455dc65069387f2dc46612f13fd45452ab301a&end=62270fb4d674fa109148c3a2618e4db9e03cd91c&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=46455dc65069387f2dc46612f13fd45452ab301a&end=62270fb4d674fa109148c3a2618e4db9e03cd91c&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=46455dc65069387f2dc46612f13fd45452ab301a&end=62270fb4d674fa109148c3a2618e4db9e03cd91c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.1% | [3.0%, 3.2%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=46455dc65069387f2dc46612f13fd45452ab301a&end=62270fb4d674fa109148c3a2618e4db9e03cd91c&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=46455dc65069387f2dc46612f13fd45452ab301a&end=62270fb4d674fa109148c3a2618e4db9e03cd91c&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 639.997s -> 639.068s (-0.15%)
**Artifact size**: 304.49 MiB -> 304.47 MiB (-0.01%)
